<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Release checklist &mdash; genomepy  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            genomepy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">Home</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/command_line.html">Command line documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/api_core.html">Python API documentation (core)</a></li>
<li class="toctree-l1"><a class="reference internal" href="_autosummary/genomepy.html">Python API documentation (full)</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/config.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/help_faq.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/about.html">About</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">genomepy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Release checklist</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/vanheeringen-lab/genomepy/blob/develop/docs/release_checklist.md" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="release-checklist">
<h1>Release checklist<a class="headerlink" href="#release-checklist" title="Permalink to this heading">ÔÉÅ</a></h1>
<ol class="arabic">
<li><p>Make sure all tests pass.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>mamba<span class="w"> </span>env<span class="w"> </span>update<span class="w"> </span>-f<span class="w"> </span>environment.yml
pytest<span class="w"> </span>-vvv
</pre></div>
</div>
</li>
<li><p>Create release candidate with <code class="docutils literal notranslate"><span class="pre">git</span> <span class="pre">flow</span></code>:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span><span class="nv">new_version</span><span class="o">=</span><span class="m">0</span>.0.0
<span class="nb">echo</span><span class="w"> </span><span class="si">${</span><span class="nv">new_version</span><span class="si">}</span>

git<span class="w"> </span>flow<span class="w"> </span>release<span class="w"> </span>start<span class="w"> </span><span class="si">${</span><span class="nv">new_version</span><span class="si">}</span>
</pre></div>
</div>
</li>
<li><p>Update version in <code class="docutils literal notranslate"><span class="pre">genomepy/__about__.py</span></code></p></li>
<li><p>Update the <code class="docutils literal notranslate"><span class="pre">CHANGELOG.md</span></code></p>
<ul class="simple">
<li><p>add the new version &amp; date to the header</p></li>
<li><p>link to the diff in the footer</p></li>
<li><p>add &amp; commit the changes, but do not push</p></li>
</ul>
</li>
<li><p>Check if release works on pypi:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>-m<span class="w"> </span>build

<span class="c1"># twine must be up to date (3.3.0 works). System installed twine can interfere.</span>
twine<span class="w"> </span>upload<span class="w"> </span>--repository-url<span class="w"> </span>https://test.pypi.org/legacy/<span class="w"> </span>dist/genomepy-<span class="si">${</span><span class="nv">new_version</span><span class="si">}</span>*

pip<span class="w"> </span>uninstall<span class="w"> </span>genomepy

<span class="c1"># the is to escape the ==, so the variable ${new_version} can be called</span>
pip<span class="w"> </span>install<span class="w"> </span>--extra-index-url<span class="w"> </span>https://test.pypi.org/simple/<span class="w"> </span>genomepy<span class="se">\=</span><span class="o">=</span><span class="si">${</span><span class="nv">new_version</span><span class="si">}</span>

<span class="c1"># tests</span>
genomepy<span class="w"> </span>--version
genomepy<span class="w"> </span>--help
genomepy<span class="w"> </span>install<span class="w"> </span>--help
genomepy<span class="w"> </span>clean
genomepy<span class="w"> </span>search<span class="w"> </span>xenopus_tropicalis
genomepy<span class="w"> </span>annotation<span class="w"> </span>hg38
genomepy<span class="w"> </span>annotation<span class="w"> </span>GRCh38.p13
genomepy<span class="w"> </span>install<span class="w"> </span>-af<span class="w"> </span>-p<span class="w"> </span>gencode<span class="w"> </span>GRCm39
genomepy<span class="w"> </span>install<span class="w"> </span>-af<span class="w"> </span>-p<span class="w"> </span>ensembl<span class="w"> </span>TAIR10
genomepy<span class="w"> </span>install<span class="w"> </span>-af<span class="w"> </span>-p<span class="w"> </span>ucsc<span class="w"> </span>sacCer3<span class="w"> </span>--UCSC-annotation<span class="w"> </span>ensGene
genomepy<span class="w"> </span>install<span class="w"> </span>-af<span class="w"> </span>-p<span class="w"> </span>ncbi<span class="w"> </span>ASM2732v1
genomepy<span class="w"> </span>install<span class="w"> </span>-af<span class="w"> </span>-p<span class="w"> </span>url<span class="w"> </span>-l<span class="w"> </span>url_test<span class="w"> </span>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/027/325/GCF_000027325.1_ASM2732v1/GCF_000027325.1_ASM2732v1_genomic.fna.gz<span class="w"> </span>--URL-to-annotation<span class="w"> </span>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/027/325/GCF_000027325.1_ASM2732v1/GCF_000027325.1_ASM2732v1_genomic.gff.gz
genomepy<span class="w"> </span>install<span class="w"> </span>-af<span class="w"> </span>-p<span class="w"> </span><span class="nb">local</span><span class="w"> </span>-l<span class="w"> </span>local_test<span class="w"> </span>~/.local/share/genomes/TAIR10/TAIR10.fa<span class="w"> </span>--Local-path-to-annotation<span class="w"> </span>~/.local/share/genomes/TAIR10/TAIR10.annotation.gtf
</pre></div>
</div>
</li>
<li><p>Finish the release:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>git<span class="w"> </span>flow<span class="w"> </span>release<span class="w"> </span>finish<span class="w"> </span><span class="si">${</span><span class="nv">new_version</span><span class="si">}</span>
</pre></div>
</div>
</li>
<li><p>Push everything to github, including tags:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>git<span class="w"> </span>push<span class="w"> </span>--follow-tags<span class="w"> </span>origin<span class="w"> </span>develop<span class="w"> </span>master
</pre></div>
</div>
</li>
<li><p>Upload to pypi:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>-m<span class="w"> </span>build
twine<span class="w"> </span>upload<span class="w"> </span>dist/genomepy-<span class="si">${</span><span class="nv">new_version</span><span class="si">}</span>*
</pre></div>
</div>
</li>
<li><p>Create release on github (if it not already exists)</p>
<ul class="simple">
<li><p>Update release with CHANGELOG information from the latest version</p></li>
<li><p>Download the tarball from the github release (<code class="docutils literal notranslate"><span class="pre">.tar.gz</span></code>).</p></li>
<li><p>Attach downloaded tarball to release as binary (this way the download count get tracked).</p></li>
</ul>
</li>
</ol>
<p>10a. Update bioconda package</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">*</span> <span class="n">wait</span> <span class="k">for</span> <span class="n">the</span> <span class="n">bioconda</span> <span class="n">bot</span> <span class="n">to</span> <span class="n">create</span> <span class="n">a</span> <span class="n">PR</span>
<span class="o">*</span> <span class="n">update</span> <span class="n">dependencies</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">bioconda</span> <span class="n">recipe</span><span class="o">.</span><span class="n">yaml</span> <span class="k">if</span> <span class="n">needed</span>
<span class="o">*</span> <span class="n">approve</span> <span class="n">the</span> <span class="n">PR</span>
<span class="o">*</span> <span class="n">comment</span><span class="p">:</span> <span class="nd">@bioconda</span><span class="o">-</span><span class="n">bot</span> <span class="n">please</span> <span class="n">merge</span>
</pre></div>
</div>
<p>10b. Update bioconda package</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>* fork bioconda/bioconda-recipes
* follow the steps in the [docs](https://bioconda.github.io/contributor/workflow.html)
* get the hash from the downloaded tarbal using `sha256sum *genomepy-${new_version}.tar.gz`
* update the [yaml file](https://github.com/bioconda/bioconda-recipes/blob/master/recipes/genomepy/meta.yaml) locally.
* push to a new branch on the fork
* start a PR
</pre></div>
</div>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright Siebren Fr√∂lich, Maarten van der Sande, Tilman Sch√§fers and Simon van Heeringen.
      <span class="lastupdated">Last updated on 2023-03-21, 14:55 (UTC).
      </span></p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>