<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Command line documentation &mdash; genomepy  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Python API documentation (core)" href="api_core.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> genomepy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../index.html">Home</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Command line documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#search-genomes-gene-annotations">Search genomes &amp; gene annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="#inspect-gene-annotations">Inspect gene annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="#install-a-genome-gene-annotation">Install a genome &amp; gene annotation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#regex-masking-compression">Regex, masking &amp; compression</a></li>
<li class="toctree-l3"><a class="reference internal" href="#other-providers-any-url-local-files">Other providers (any URL/local files)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#reproducibility">Reproducibility</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#plugins">Plugins</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#genome-blacklists">Genome blacklists</a></li>
<li class="toctree-l3"><a class="reference internal" href="#aligner-indexes">Aligner indexes</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id1">Configuration</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#genome-location">Genome location</a></li>
<li class="toctree-l3"><a class="reference internal" href="#compression">Compression</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#list-available-providers">List available providers</a></li>
<li class="toctree-l2"><a class="reference internal" href="#list-available-genomes">List available genomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="#local-cache">Local cache.</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="api_core.html">Python API documentation (core)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../_autosummary/genomepy.html">Python API documentation (full)</a></li>
<li class="toctree-l1"><a class="reference internal" href="help_faq.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="about.html">About</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">genomepy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">Command line documentation</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/vanheeringen-lab/genomepy/blob/develop/docs/content/command_line.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="command-line-documentation">
<h1>Command line documentation<a class="headerlink" href="#command-line-documentation" title="Permalink to this heading"></a></h1>
<p>All commands come with a short explanation when appended with <code class="docutils literal notranslate"><span class="pre">-h</span></code>/<code class="docutils literal notranslate"><span class="pre">--help</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>genomepy<span class="w"> </span>--help
Usage:<span class="w"> </span>genomepy<span class="w"> </span><span class="o">[</span>OPTIONS<span class="o">]</span><span class="w"> </span>COMMAND<span class="w"> </span><span class="o">[</span>ARGS<span class="o">]</span>...

Options:
<span class="w">  </span>--version<span class="w">   </span>Show<span class="w"> </span>the<span class="w"> </span>version<span class="w"> </span>and<span class="w"> </span>exit.
<span class="w">  </span>-h,<span class="w"> </span>--help<span class="w">  </span>Show<span class="w"> </span>this<span class="w"> </span>message<span class="w"> </span>and<span class="w"> </span>exit.

Commands:
<span class="w">  </span>annotation<span class="w">  </span>show<span class="w"> </span>1st<span class="w"> </span>lines<span class="w"> </span>of<span class="w"> </span>each<span class="w"> </span>annotation
<span class="w">  </span>clean<span class="w">       </span>remove<span class="w"> </span>provider<span class="w"> </span>data
<span class="w">  </span>config<span class="w">      </span>manage<span class="w"> </span>configuration
<span class="w">  </span>genomes<span class="w">     </span>list<span class="w"> </span>available<span class="w"> </span>genomes
<span class="w">  </span>install<span class="w">     </span>install<span class="w"> </span>a<span class="w"> </span>genome<span class="w"> </span><span class="p">&amp;</span><span class="w"> </span>run<span class="w"> </span>active<span class="w"> </span>plugins
<span class="w">  </span>plugin<span class="w">      </span>manage<span class="w"> </span>plugins
<span class="w">  </span>providers<span class="w">   </span>list<span class="w"> </span>available<span class="w"> </span>providers
<span class="w">  </span>search<span class="w">      </span>search<span class="w"> </span><span class="k">for</span><span class="w"> </span>genomes
</pre></div>
</div>
<section id="search-genomes-gene-annotations">
<h2>Search genomes &amp; gene annotations<a class="headerlink" href="#search-genomes-gene-annotations" title="Permalink to this heading"></a></h2>
<p>Let’s say we want to download a <em>Xenopus tropicalis</em> genome &amp; gene annotation.
First, lets find out what’s out there!</p>
<p>You can search by name, taxonomy ID or assembly accession ID.
Additionally, you can limit the search result to one provider with <code class="docutils literal notranslate"><span class="pre">-p</span></code>/<code class="docutils literal notranslate"><span class="pre">--provider</span></code>.
Furthermore, you can get the absolute <code class="docutils literal notranslate"><span class="pre">--size</span></code> of each genome (this option slows down the search).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>genomepy<span class="w"> </span>search<span class="w"> </span>xenopus<span class="w"> </span>tro
name<span class="w">                       </span>provider<span class="w">    </span>accession<span class="w">           </span>tax_id<span class="w">    </span>annotation<span class="w">     </span>species<span class="w">               </span>other_info
<span class="w">                                                                      </span>n<span class="w"> </span>r<span class="w"> </span>e<span class="w"> </span>k
Xenopus_tropicalis_v9.1<span class="w">    </span>Ensembl<span class="w">     </span>GCA_000004195.3<span class="w">       </span><span class="m">8364</span><span class="w">        </span>✓<span class="w">          </span>Xenopus<span class="w"> </span>tropicalis<span class="w">    </span><span class="m">2019</span>-04-Ensembl/2019-12
xenTro1<span class="w">                    </span>UCSC<span class="w">        </span>na<span class="w">                    </span><span class="m">8364</span><span class="w">     </span>✗<span class="w"> </span>✗<span class="w"> </span>✗<span class="w"> </span>✗<span class="w">       </span>Xenopus<span class="w"> </span>tropicalis<span class="w">    </span>Oct.<span class="w"> </span><span class="m">2004</span><span class="w"> </span><span class="o">(</span>JGI<span class="w"> </span><span class="m">3</span>.0/xenTro1<span class="o">)</span>
xenTro2<span class="w">                    </span>UCSC<span class="w">        </span>na<span class="w">                    </span><span class="m">8364</span><span class="w">     </span>✗<span class="w"> </span>✓<span class="w"> </span>✓<span class="w"> </span>✗<span class="w">       </span>Xenopus<span class="w"> </span>tropicalis<span class="w">    </span>Aug.<span class="w"> </span><span class="m">2005</span><span class="w"> </span><span class="o">(</span>JGI<span class="w"> </span><span class="m">4</span>.1/xenTro2<span class="o">)</span>
xenTro3<span class="w">                    </span>UCSC<span class="w">        </span>GCA_000004195.1<span class="w">       </span><span class="m">8364</span><span class="w">     </span>✗<span class="w"> </span>✓<span class="w"> </span>✓<span class="w"> </span>✗<span class="w">       </span>Xenopus<span class="w"> </span>tropicalis<span class="w">    </span>Nov.<span class="w"> </span><span class="m">2009</span><span class="w"> </span><span class="o">(</span>JGI<span class="w"> </span><span class="m">4</span>.2/xenTro3<span class="o">)</span>
xenTro7<span class="w">                    </span>UCSC<span class="w">        </span>GCA_000004195.2<span class="w">       </span><span class="m">8364</span><span class="w">     </span>✓<span class="w"> </span>✓<span class="w"> </span>✗<span class="w"> </span>✗<span class="w">       </span>Xenopus<span class="w"> </span>tropicalis<span class="w">    </span>Sep.<span class="w"> </span><span class="m">2012</span><span class="w"> </span><span class="o">(</span>JGI<span class="w"> </span><span class="m">7</span>.0/xenTro7<span class="o">)</span>
xenTro9<span class="w">                    </span>UCSC<span class="w">        </span>GCA_000004195.3<span class="w">       </span><span class="m">8364</span><span class="w">     </span>✓<span class="w"> </span>✓<span class="w"> </span>✓<span class="w"> </span>✗<span class="w">       </span>Xenopus<span class="w"> </span>tropicalis<span class="w">    </span>Jul.<span class="w"> </span><span class="m">2016</span><span class="w"> </span><span class="o">(</span>Xenopus_tropicalis_v9.1/xenTro9<span class="o">)</span>
Xtropicalis_v7<span class="w">             </span>NCBI<span class="w">        </span>GCF_000004195.2<span class="w">       </span><span class="m">8364</span><span class="w">        </span>✓<span class="w">          </span>Xenopus<span class="w"> </span>tropicalis<span class="w">    </span>DOE<span class="w"> </span>Joint<span class="w"> </span>Genome<span class="w"> </span>Institute
Xenopus_tropicalis_v9.1<span class="w">    </span>NCBI<span class="w">        </span>GCF_000004195.3<span class="w">       </span><span class="m">8364</span><span class="w">        </span>✓<span class="w">          </span>Xenopus<span class="w"> </span>tropicalis<span class="w">    </span>DOE<span class="w"> </span>Joint<span class="w"> </span>Genome<span class="w"> </span>Institute
UCB_Xtro_10.0<span class="w">              </span>NCBI<span class="w">        </span>GCF_000004195.4<span class="w">       </span><span class="m">8364</span><span class="w">        </span>✓<span class="w">          </span>Xenopus<span class="w"> </span>tropicalis<span class="w">    </span>University<span class="w"> </span>of<span class="w"> </span>California,<span class="w"> </span>Berkeley
ASM1336827v1<span class="w">               </span>NCBI<span class="w">        </span>GCA_013368275.1<span class="w">       </span><span class="m">8364</span><span class="w">        </span>✗<span class="w">          </span>Xenopus<span class="w"> </span>tropicalis<span class="w">    </span>Southern<span class="w"> </span>University<span class="w"> </span>of<span class="w"> </span>Science<span class="w"> </span>and<span class="w"> </span>Technology
<span class="w"> </span>^
<span class="w"> </span>Use<span class="w"> </span>name<span class="w"> </span><span class="k">for</span><span class="w"> </span>genomepy<span class="w"> </span>install
</pre></div>
</div>
</section>
<section id="inspect-gene-annotations">
<h2>Inspect gene annotations<a class="headerlink" href="#inspect-gene-annotations" title="Permalink to this heading"></a></h2>
<p>Let’s say we want to download the <em>Xenopus tropicalis</em> genome &amp; gene annotation from UCSC.</p>
<p>Since we are interested in the gene annotation as well, we should check which gene annotation suits our needs.
As you can see in the search results, UCSC has several gene annotations for us to choose from.
In the search results, <code class="docutils literal notranslate"><span class="pre">n</span> <span class="pre">r</span> <span class="pre">e</span> <span class="pre">k</span></code> denotes which UCSC annotations are available.
These stand for <strong>n</strong>cbiRefSeq, <strong>r</strong>efGene, <strong>e</strong>nsGene and <strong>k</strong>nownGene, respectively.</p>
<p>We can quickly inspect these with the <code class="docutils literal notranslate"><span class="pre">genomepy</span> <span class="pre">annotation</span></code> command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>genomepy<span class="w"> </span>annotation<span class="w"> </span>xenTro9<span class="w"> </span>-p<span class="w"> </span>ucsc
<span class="m">12</span>:04:41<span class="w"> </span><span class="p">|</span><span class="w"> </span>INFO<span class="w"> </span><span class="p">|</span><span class="w"> </span>UCSC<span class="w"> </span>ncbiRefSeq
chr1<span class="w">    </span>genomepy<span class="w">        </span>transcript<span class="w">      </span><span class="m">133270</span><span class="w">  </span><span class="m">152620</span><span class="w">  </span>.<span class="w">       </span>-<span class="w">       </span>.<span class="w">       </span>gene_id<span class="w"> </span><span class="s2">&quot;LOC100490505&quot;</span><span class="p">;</span><span class="w"> </span>transcript_id<span class="w"> </span><span class="s2">&quot;XM_012956089.1&quot;</span><span class="p">;</span><span class="w">  </span>gene_name<span class="w"> </span><span class="s2">&quot;LOC100490505&quot;</span><span class="p">;</span>
chr1<span class="w">    </span>genomepy<span class="w">        </span>exon<span class="w">    </span><span class="m">133270</span><span class="w">  </span><span class="m">134186</span><span class="w">  </span>.<span class="w">       </span>-<span class="w">       </span>.<span class="w">       </span>gene_id<span class="w"> </span><span class="s2">&quot;LOC100490505&quot;</span><span class="p">;</span><span class="w"> </span>transcript_id<span class="w"> </span><span class="s2">&quot;XM_012956089.1&quot;</span><span class="p">;</span><span class="w"> </span>exon_number<span class="w"> </span><span class="s2">&quot;1&quot;</span><span class="p">;</span><span class="w"> </span>exon_id<span class="w"> </span><span class="s2">&quot;XM_012956089.1.1&quot;</span><span class="p">;</span><span class="w"> </span>gene_name<span class="w"> </span><span class="s2">&quot;LOC100490505&quot;</span><span class="p">;</span>
<span class="m">12</span>:04:45<span class="w"> </span><span class="p">|</span><span class="w"> </span>INFO<span class="w"> </span><span class="p">|</span><span class="w"> </span>UCSC<span class="w"> </span>refGene
chr1<span class="w">    </span>genomepy<span class="w">        </span>transcript<span class="w">      </span><span class="m">193109390</span><span class="w">       </span><span class="m">193134311</span><span class="w">       </span>.<span class="w">       </span>+<span class="w">       </span>.<span class="w">       </span>gene_id<span class="w"> </span><span class="s2">&quot;pias2&quot;</span><span class="p">;</span><span class="w"> </span>transcript_id<span class="w"> </span><span class="s2">&quot;NM_001078987&quot;</span><span class="p">;</span><span class="w">  </span>gene_name<span class="w"> </span><span class="s2">&quot;pias2&quot;</span><span class="p">;</span>
chr1<span class="w">    </span>genomepy<span class="w">        </span>exon<span class="w">    </span><span class="m">193109390</span><span class="w">       </span><span class="m">193109458</span><span class="w">       </span>.<span class="w">       </span>+<span class="w">       </span>.<span class="w">       </span>gene_id<span class="w"> </span><span class="s2">&quot;pias2&quot;</span><span class="p">;</span><span class="w"> </span>transcript_id<span class="w"> </span><span class="s2">&quot;NM_001078987&quot;</span><span class="p">;</span><span class="w"> </span>exon_number<span class="w"> </span><span class="s2">&quot;1&quot;</span><span class="p">;</span><span class="w"> </span>exon_id<span class="w"> </span><span class="s2">&quot;NM_001078987.1&quot;</span><span class="p">;</span><span class="w"> </span>gene_name<span class="w"> </span><span class="s2">&quot;pias2&quot;</span><span class="p">;</span>
<span class="m">12</span>:04:49<span class="w"> </span><span class="p">|</span><span class="w"> </span>INFO<span class="w"> </span><span class="p">|</span><span class="w"> </span>UCSC<span class="w"> </span>ensGene
chr1<span class="w">    </span>genomepy<span class="w">        </span>transcript<span class="w">      </span><span class="m">133270</span><span class="w">  </span><span class="m">152620</span><span class="w">  </span>.<span class="w">       </span>-<span class="w">       </span>.<span class="w">       </span>gene_id<span class="w"> </span><span class="s2">&quot;ENSXETG00000030302.2&quot;</span><span class="p">;</span><span class="w"> </span>transcript_id<span class="w"> </span><span class="s2">&quot;ENSXETT00000061673.2&quot;</span><span class="p">;</span><span class="w">  </span>gene_name<span class="w"> </span><span class="s2">&quot;ENSXETG00000030302.2&quot;</span><span class="p">;</span>
chr1<span class="w">    </span>genomepy<span class="w">        </span>exon<span class="w">    </span><span class="m">133270</span><span class="w">  </span><span class="m">134186</span><span class="w">  </span>.<span class="w">       </span>-<span class="w">       </span>.<span class="w">       </span>gene_id<span class="w"> </span><span class="s2">&quot;ENSXETG00000030302.2&quot;</span><span class="p">;</span><span class="w"> </span>transcript_id<span class="w"> </span><span class="s2">&quot;ENSXETT00000061673.2&quot;</span><span class="p">;</span><span class="w"> </span>exon_number<span class="w"> </span><span class="s2">&quot;1&quot;</span><span class="p">;</span><span class="w"> </span>exon_id<span class="w"> </span><span class="s2">&quot;ENSXETT00000061673.2.1&quot;</span><span class="p">;</span><span class="w"> </span>gene_name<span class="w"> </span><span class="s2">&quot;ENSXETG00000030302.2&quot;</span><span class="p">;</span>
</pre></div>
</div>
<p>Here we can see that the <code class="docutils literal notranslate"><span class="pre">refGene</span></code> annotation has actual HGNC gene names, so lets go with this annotation.
This differs between assemblies, so be sure to check!</p>
</section>
<section id="install-a-genome-gene-annotation">
<h2>Install a genome &amp; gene annotation<a class="headerlink" href="#install-a-genome-gene-annotation" title="Permalink to this heading"></a></h2>
<p>Copy the name returned by the search function to install.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>genomepy<span class="w"> </span>install<span class="w"> </span>xenTro9
</pre></div>
</div>
<p>You can choose to download gene annotation files with the <code class="docutils literal notranslate"><span class="pre">-a</span></code>/<code class="docutils literal notranslate"><span class="pre">--annotation</span></code> option.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>genomepy<span class="w"> </span>install<span class="w"> </span>xenTro9<span class="w"> </span>--annotation
</pre></div>
</div>
<p>For UCSC we can also select the annotation type.
See <code class="docutils literal notranslate"><span class="pre">genomepy</span> <span class="pre">install</span> <span class="pre">--help</span></code> for all provider specific options.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>genomepy<span class="w"> </span>install<span class="w"> </span>xenTro9<span class="w"> </span>--UCSC-annotation<span class="w"> </span>refGene
</pre></div>
</div>
<p>Since we did not specify the provider here, genomepy will use the first provider with <code class="docutils literal notranslate"><span class="pre">xenTro9</span></code>.
You can specify a provider by name with <code class="docutils literal notranslate"><span class="pre">-p</span></code>/<code class="docutils literal notranslate"><span class="pre">--provider</span></code>:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>genomepy<span class="w"> </span>install<span class="w"> </span>xenTro9<span class="w"> </span>-p<span class="w"> </span>UCSC
Downloading<span class="w"> </span>genome<span class="w"> </span>from<span class="w"> </span>http://hgdownload.soe.ucsc.edu/goldenPath/xenTro9/bigZips/xenTro9.fa.gz...
Genome<span class="w"> </span>download<span class="w"> </span>successful,<span class="w"> </span>starting<span class="w"> </span>post<span class="w"> </span>processing...

name:<span class="w"> </span>xenTro9
<span class="nb">local</span><span class="w"> </span>name:<span class="w"> </span>xenTro9
fasta:<span class="w"> </span>~/.local/share/genomes/xenTro9/xenTro9.fa
</pre></div>
</div>
<p>Next, the genome is downloaded to the directory specified in the config file (by default <code class="docutils literal notranslate"><span class="pre">~/.local/share/genomes</span></code>).
To choose a different directory, use the <code class="docutils literal notranslate"><span class="pre">-g</span></code>/<code class="docutils literal notranslate"><span class="pre">--genomes_dir</span></code> option:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>genomepy<span class="w"> </span>install<span class="w"> </span>sacCer3<span class="w"> </span>-p<span class="w"> </span>UCSC<span class="w"> </span>-g<span class="w"> </span>/path/to/my/genomes
Downloading<span class="w"> </span>genome<span class="w"> </span>from<span class="w"> </span>http://hgdownload.soe.ucsc.edu/goldenPath/sacCer3/bigZips/chromFa.tar.gz...
Genome<span class="w"> </span>download<span class="w"> </span>successful,<span class="w"> </span>starting<span class="w"> </span>post<span class="w"> </span>processing...

name:<span class="w"> </span>sacCer3
<span class="nb">local</span><span class="w"> </span>name:<span class="w"> </span>sacCer3
fasta:<span class="w"> </span>/path/to/my/genomes/sacCer3/sacCer3.fa
</pre></div>
</div>
<section id="regex-masking-compression">
<h3>Regex, masking &amp; compression<a class="headerlink" href="#regex-masking-compression" title="Permalink to this heading"></a></h3>
<p>You can use a regular expression to filter for matching sequences
(or non-matching sequences by using the <code class="docutils literal notranslate"><span class="pre">-n</span></code>/<code class="docutils literal notranslate"><span class="pre">--no-match</span></code> option).
For instance, the following command downloads hg38 and saves only the major chromosomes:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>genomepy<span class="w"> </span>install<span class="w"> </span>hg38<span class="w"> </span>-p<span class="w"> </span>UCSC<span class="w"> </span>-r<span class="w"> </span><span class="s1">&#39;chr[0-9XY]+$&#39;</span>
Downloading<span class="w"> </span>genome<span class="w"> </span>from<span class="w"> </span>from<span class="w"> </span>http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz...
Genome<span class="w"> </span>download<span class="w"> </span>successful,<span class="w"> </span>starting<span class="w"> </span>post<span class="w"> </span>processing...

name:<span class="w"> </span>hg38
<span class="nb">local</span><span class="w"> </span>name:<span class="w"> </span>hg38
fasta:<span class="w"> </span>/data/genomes/hg38/hg38.fa

$<span class="w"> </span>grep<span class="w"> </span><span class="s2">&quot;&gt;&quot;</span><span class="w"> </span>/data/genomes/hg38/hg38.fa
&gt;chr1
&gt;chr10
&gt;chr11
&gt;chr12
&gt;chr13
&gt;chr14
&gt;chr15
&gt;chr16
&gt;chr17
&gt;chr18
&gt;chr19
&gt;chr2
&gt;chr20
&gt;chr21
&gt;chr22
&gt;chr3
&gt;chr4
&gt;chr5
&gt;chr6
&gt;chr7
&gt;chr8
&gt;chr9
&gt;chrX
&gt;chrY
</pre></div>
</div>
<p>By default, genome sequences are soft-masked (ACgtN).
Use <code class="docutils literal notranslate"><span class="pre">-m</span> <span class="pre">hard</span></code> for hard masking (ACNNN), or <code class="docutils literal notranslate"><span class="pre">-m</span> <span class="pre">none</span></code> for no masking (ACGTN).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>genomepy<span class="w"> </span>install<span class="w"> </span>hg38<span class="w"> </span>--mask<span class="w"> </span>hard
</pre></div>
</div>
<p>If you wish to conserve space, you can tell genomepy to compress the downloaded data by passing the <code class="docutils literal notranslate"><span class="pre">-b</span></code>/<code class="docutils literal notranslate"><span class="pre">--bgzip</span></code> option.
See <a class="reference external" href="#compression">Configuration</a> for details.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>genomepy<span class="w"> </span>install<span class="w"> </span>hg38<span class="w"> </span>--bgzip
</pre></div>
</div>
</section>
<section id="other-providers-any-url-local-files">
<h3>Other providers (any URL/local files)<a class="headerlink" href="#other-providers-any-url-local-files" title="Permalink to this heading"></a></h3>
<p>To use assemblies not on NCBI, UCSC, Ensembl or GENCODE, you can give a URL instead of a name, together with <code class="docutils literal notranslate"><span class="pre">--provider</span> <span class="pre">URL</span></code>.
Similarly, if you have a local FASTA file, you can install this using the filepath, together with <code class="docutils literal notranslate"><span class="pre">--provider</span> <span class="pre">Local</span></code>:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>genomepy<span class="w"> </span>install<span class="w"> </span>-p<span class="w"> </span>url<span class="w"> </span>https://research.nhgri.nih.gov/hydra/download/assembly/<span class="se">\H</span>m105_Dovetail_Assembly_1.0.fa.gz
</pre></div>
</div>
<p>This will install the genome under the filename of the URL/filepath, but can be changed with the <code class="docutils literal notranslate"><span class="pre">-l</span></code>/<code class="docutils literal notranslate"><span class="pre">--localname</span></code> option.</p>
<p>If you add the <code class="docutils literal notranslate"><span class="pre">--annotation</span></code> flag, genomepy will search the (remote) directory for an annotation file as well.
Should this fail, you can also add a URL to the annotation with <code class="docutils literal notranslate"><span class="pre">--URL-to-annotation</span></code> with the <code class="docutils literal notranslate"><span class="pre">URL</span></code> provider,
or a filepath with <code class="docutils literal notranslate"><span class="pre">--Local-path-to-annotation</span></code> with the <code class="docutils literal notranslate"><span class="pre">Local</span></code> provider:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>genomepy<span class="w"> </span>install<span class="w"> </span>-p<span class="w"> </span><span class="nb">local</span><span class="w"> </span>/path/to/genome.fa<span class="w"> </span>--Local-path-to-annotation<span class="w"> </span>/path/to/gene.annotation.gtf
</pre></div>
</div>
</section>
<section id="reproducibility">
<h3>Reproducibility<a class="headerlink" href="#reproducibility" title="Permalink to this heading"></a></h3>
<p>All selected options are stored in a <code class="docutils literal notranslate"><span class="pre">README.txt</span></code>.
This includes the original name, download location and other genomepy operations (such as regex filtering and time).</p>
</section>
</section>
<section id="plugins">
<h2>Plugins<a class="headerlink" href="#plugins" title="Permalink to this heading"></a></h2>
<p>Plugins are optional steps that are executed after installing an assembly with <code class="docutils literal notranslate"><span class="pre">genomepy</span> <span class="pre">install</span></code>.
If you already installed an assembly, you can activate a plugin and rerun the install command.
This will not overwrite your local files, unless you use the <code class="docutils literal notranslate"><span class="pre">--force</span></code> option.</p>
<p>Check which plugins are enabled with <code class="docutils literal notranslate"><span class="pre">genomepy</span> <span class="pre">plugin</span> <span class="pre">list</span></code>.</p>
<section id="genome-blacklists">
<h3>Genome blacklists<a class="headerlink" href="#genome-blacklists" title="Permalink to this heading"></a></h3>
<p>For some model organisms, genomepy can download a genome blacklist (generated by the <a class="reference external" href="https://www.nature.com/articles/s41598-019-45839-z">Kundaje lab</a>).
Blacklists are only available for these model organisms when downloaded from UCSC, and for the human and mouse genomes.</p>
<p>Enable the blacklist plugin to use it:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>genomepy<span class="w"> </span>plugin<span class="w"> </span><span class="nb">enable</span><span class="w"> </span>blacklist
Enabled<span class="w"> </span>plugins:<span class="w"> </span>blacklist
</pre></div>
</div>
</section>
<section id="aligner-indexes">
<h3>Aligner indexes<a class="headerlink" href="#aligner-indexes" title="Permalink to this heading"></a></h3>
<p>You can also create aligner indexes for several widely used aligners.
Currently, genomepy supports:</p>
<ul class="simple">
<li><p><a class="reference external" href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">Bowtie2</a></p></li>
<li><p><a class="reference external" href="http://bio-bwa.sourceforge.net/">BWA</a></p></li>
<li><p><a class="reference external" href="http://research-pub.gene.com/gmap/">GMAP</a></p></li>
<li><p><a class="reference external" href="https://ccb.jhu.edu/software/hisat2/index.shtml">HISAT2</a></p></li>
<li><p><a class="reference external" href="https://github.com/lh3/minimap2">Minimap2</a></p></li>
<li><p><a class="reference external" href="https://github.com/alexdobin/STAR">STAR</a></p></li>
</ul>
<p>These programs are not installed by genomepy and need to be installed separately for the indexing to work.
The easiest way to do so is with conda, e.g.: <code class="docutils literal notranslate"><span class="pre">conda</span> <span class="pre">install</span> <span class="pre">-c</span> <span class="pre">bioconda</span> <span class="pre">bwa</span> <span class="pre">star</span></code></p>
<p>Splice-aware indexing (required for e.g. RNA-seq) can be performed by STAR and Hisat2.
This will be done automatically if the gene annotation was downloaded as well.
Finally, STAR can further improve mapping to (novel) splice junctions by indexing again (see 2-pass mapping mode in the STAR manual).
The second pass is not supported by genomepy.</p>
<p>You can configure the index creation with <code class="docutils literal notranslate"><span class="pre">genomepy</span> <span class="pre">plugin</span> <span class="pre">enable</span></code>, e.g.:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>genomepy<span class="w"> </span>plugin<span class="w"> </span><span class="nb">enable</span><span class="w"> </span>bwa<span class="w"> </span>star
Enabled<span class="w"> </span>plugins:<span class="w"> </span>blacklist,<span class="w"> </span>bwa,<span class="w"> </span>star
</pre></div>
</div>
<p>You can pass the number of threads to use for aligner index creation with <code class="docutils literal notranslate"><span class="pre">genomepy</span> <span class="pre">install</span> <span class="pre">--threads</span></code> (default is 8).</p>
</section>
</section>
<section id="id1">
<h2>Configuration<a class="headerlink" href="#id1" title="Permalink to this heading"></a></h2>
<p>All defaults can be overwritten on the command line and in Python.
However, you can create &amp; edit the config file to change the default settings:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>genomepy<span class="w"> </span>config<span class="w"> </span>generate
Created<span class="w"> </span>config<span class="w"> </span>file<span class="w"> </span>~/.config/genomepy/genomepy.yaml
</pre></div>
</div>
<section id="genome-location">
<h3>Genome location<a class="headerlink" href="#genome-location" title="Permalink to this heading"></a></h3>
<p>By default, genomes will be saved in <code class="docutils literal notranslate"><span class="pre">~/.local/share/genomes</span></code>.</p>
<p>To set the default genome directory, to <code class="docutils literal notranslate"><span class="pre">/data/genomes</span></code> for instance,
edit <code class="docutils literal notranslate"><span class="pre">~/.config/genomepy/genomepy.yaml</span></code> and change the following line:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">genomes_dir</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">/data/genomes</span>
</pre></div>
</div>
</section>
<section id="compression">
<h3>Compression<a class="headerlink" href="#compression" title="Permalink to this heading"></a></h3>
<p>Genome FASTA files can be stored using bgzip compression.
This means that the FASTA files will take up less space on disk.
Set the following line to your config file:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">bgzip</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
</pre></div>
</div>
<p>Most tools are able to use bgzip-compressed genome files.
One notable exception is <code class="docutils literal notranslate"><span class="pre">bedtools</span> <span class="pre">getfasta</span></code>.
As an alternative, you can use the <code class="docutils literal notranslate"><span class="pre">faidx</span></code> command-line script from <a class="reference external" href="https://github.com/mdshw5/pyfaidx">pyfaidx</a>
which comes installed with genomepy.</p>
</section>
</section>
<section id="list-available-providers">
<h2>List available providers<a class="headerlink" href="#list-available-providers" title="Permalink to this heading"></a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>genomepy<span class="w"> </span>providers
GENCODE
Ensembl
UCSC
NCBI
Local
URL
</pre></div>
</div>
</section>
<section id="list-available-genomes">
<h2>List available genomes<a class="headerlink" href="#list-available-genomes" title="Permalink to this heading"></a></h2>
<p>You can constrain the genome list by using the <code class="docutils literal notranslate"><span class="pre">-p</span></code>/<code class="docutils literal notranslate"><span class="pre">--provider</span></code> option to search only a specific provider.
Additionally, you can get the absolute <code class="docutils literal notranslate"><span class="pre">--size</span></code> of each genome (this option slows down the search).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>genomepy<span class="w"> </span>genomes<span class="w"> </span>-p<span class="w"> </span>UCSC
name<span class="w">                    </span>provider<span class="w">    </span>accession<span class="w">          </span>tax_id<span class="w">     </span>annotation<span class="w">     </span>species<span class="w">                                     </span>other_info
<span class="w">                                                                   </span>n<span class="w"> </span>r<span class="w"> </span>e<span class="w"> </span>k
ailMel1<span class="w">                 </span>UCSC<span class="w">        </span>GCF_000004335.2<span class="w">      </span><span class="m">9646</span><span class="w">      </span>✓<span class="w"> </span>✗<span class="w"> </span>✓<span class="w"> </span>✗<span class="w">       </span>Ailuropoda<span class="w"> </span>melanoleuca<span class="w">                      </span>Dec.<span class="w"> </span><span class="m">2009</span><span class="w"> </span><span class="o">(</span>BGI-Shenzhen<span class="w"> </span><span class="m">1</span>.0/ailMel1<span class="o">)</span>
allMis1<span class="w">                 </span>UCSC<span class="w">        </span>GCA_000281125.1<span class="w">      </span><span class="m">8496</span><span class="w">      </span>✗<span class="w"> </span>✓<span class="w"> </span>✗<span class="w"> </span>✗<span class="w">       </span>Alligator<span class="w"> </span>mississippiensis<span class="w">                  </span>Aug.<span class="w"> </span><span class="m">2012</span><span class="w"> </span><span class="o">(</span>allMis0.2/allMis1<span class="o">)</span>
anoCar1<span class="w">                 </span>UCSC<span class="w">        </span>na<span class="w">                  </span><span class="m">28377</span><span class="w">      </span>✗<span class="w"> </span>✗<span class="w"> </span>✓<span class="w"> </span>✗<span class="w">       </span>Anolis<span class="w"> </span>carolinensis<span class="w">                         </span>Feb.<span class="w"> </span><span class="m">2007</span><span class="w"> </span><span class="o">(</span>Broad/anoCar1<span class="o">)</span>
</pre></div>
</div>
</section>
<section id="local-cache">
<h2>Local cache.<a class="headerlink" href="#local-cache" title="Permalink to this heading"></a></h2>
<p>Note that the first time you run <code class="docutils literal notranslate"><span class="pre">genomepy</span> <span class="pre">search</span></code> or <code class="docutils literal notranslate"><span class="pre">list</span></code>
the command will take a while as the genome lists have to be downloaded.
The lists are cached locally, which will save time later.
The cached files are stored in <code class="docutils literal notranslate"><span class="pre">~/.cache/genomepy</span></code> and expire after 7 days.
You can also delete this directory to clean the cache using <code class="docutils literal notranslate"><span class="pre">genomepy</span> <span class="pre">clean</span></code>.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="api_core.html" class="btn btn-neutral float-right" title="Python API documentation (core)" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright Siebren Frölich, Maarten van der Sande, Tilman Schäfers and Simon van Heeringen.
      <span class="lastupdated">Last updated on 2023-01-13, 15:18 (UTC).
      </span></p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>