<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>genomepy.functions.install_genome &mdash; genomepy  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="genomepy.functions.list_available_genomes" href="genomepy.functions.list_available_genomes.html" />
    <link rel="prev" title="genomepy.functions.head_annotations" href="genomepy.functions.head_annotations.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            genomepy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../index.html">Home</a></li>
<li class="toctree-l1"><a class="reference internal" href="../content/installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../content/command_line.html">Command line documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../content/api_core.html">Python API documentation (core)</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="genomepy.html">Python API documentation (full)</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="genomepy.annotation.html">genomepy.annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.argparse_support.html">genomepy.argparse_support</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.caching.html">genomepy.caching</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.cli.html">genomepy.cli</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.config.html">genomepy.config</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.exceptions.html">genomepy.exceptions</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.files.html">genomepy.files</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="genomepy.functions.html">genomepy.functions</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="genomepy.functions.generate_env.html">genomepy.functions.generate_env</a></li>
<li class="toctree-l3"><a class="reference internal" href="genomepy.functions.head_annotations.html">genomepy.functions.head_annotations</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">genomepy.functions.install_genome</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genomepy.functions.install_genome"><code class="docutils literal notranslate"><span class="pre">install_genome()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="genomepy.functions.list_available_genomes.html">genomepy.functions.list_available_genomes</a></li>
<li class="toctree-l3"><a class="reference internal" href="genomepy.functions.list_installed_genomes.html">genomepy.functions.list_installed_genomes</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.genome.html">genomepy.genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.online.html">genomepy.online</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.plugins.html">genomepy.plugins</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.providers.html">genomepy.providers</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.seq.html">genomepy.seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.utils.html">genomepy.utils</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../content/config.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../content/help_faq.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="../content/about.html">About</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">genomepy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="genomepy.html">genomepy</a></li>
          <li class="breadcrumb-item"><a href="genomepy.functions.html">genomepy.functions</a></li>
      <li class="breadcrumb-item active">genomepy.functions.install_genome</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/vanheeringen-lab/genomepy/blob/develop/docs/_autosummary/genomepy.functions.install_genome.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="genomepy-functions-install-genome">
<h1>genomepy.functions.install_genome<a class="headerlink" href="#genomepy-functions-install-genome" title="Permalink to this heading"></a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="genomepy.functions.install_genome">
<span class="sig-prename descclassname"><span class="pre">genomepy.functions.</span></span><span class="sig-name descname"><span class="pre">install_genome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">provider</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genomes_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">localname</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'soft'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_alt</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regex</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">invert_match</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bgzip</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">only_annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip_matching</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip_filter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">dict</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="genomepy.genome.Genome.html#genomepy.genome.Genome" title="genomepy.genome.Genome"><span class="pre">Genome</span></a></span></span><a class="headerlink" href="#genomepy.functions.install_genome" title="Permalink to this definition"></a></dt>
<dd><p>Install a genome (&amp; gene annotation).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – Genome name</p></li>
<li><p><strong>provider</strong> (<em>str</em><em> , </em><em>optional</em>) – Provider name. will try Gencode, Ensembl, UCSC and NCBI (in that order) if not specified.</p></li>
<li><p><strong>genomes_dir</strong> (<em>str</em><em> , </em><em>optional</em>) – Where to create the output folder.</p></li>
<li><p><strong>localname</strong> (<em>str</em><em> , </em><em>optional</em>) – Custom name for this genome.</p></li>
<li><p><strong>mask</strong> (<em>str</em><em> , </em><em>optional</em>) – Genome masking of repetitive sequences. Options: hard/soft/none, default is soft.</p></li>
<li><p><strong>keep_alt</strong> (<em>bool</em><em> , </em><em>optional</em>) – Some genomes contain alternative regions. These regions cause issues with
sequence alignment, as they are inherently duplications of the consensus regions.
Set to true to keep these alternative regions.</p></li>
<li><p><strong>regex</strong> (<em>str</em><em> , </em><em>optional</em>) – Regular expression to select specific chromosome / scaffold names.</p></li>
<li><p><strong>invert_match</strong> (<em>bool</em><em> , </em><em>optional</em>) – Set to True to select all chromosomes that <em>don’t</em> match the regex.</p></li>
<li><p><strong>bgzip</strong> (<em>bool</em><em> , </em><em>optional</em>) – If set to True the genome FASTA file will be compressed using bgzip,
and gene annotation will be compressed with gzip.</p></li>
<li><p><strong>threads</strong> (<em>int</em><em> , </em><em>optional</em>) – Build genome index using multithreading (if supported). Default: lowest of 8/all threads.</p></li>
<li><p><strong>force</strong> (<em>bool</em><em> , </em><em>optional</em>) – Set to True to overwrite existing files.</p></li>
<li><p><strong>annotation</strong> (<em>bool</em><em> , </em><em>optional</em>) – If set to True, download gene annotation in BED and GTF format.</p></li>
<li><p><strong>only_annotation</strong> (<em>bool</em><em> , </em><em>optional</em>) – If set to True, only download the gene annotation files.</p></li>
<li><p><strong>skip_matching</strong> (<em>bool</em><em> , </em><em>optional</em>) – If set to True, contigs in the annotation not matching
those in the genome will not be corrected.</p></li>
<li><p><strong>skip_filter</strong> (<em>bool</em><em> , </em><em>optional</em>) – If set to True, the gene annotations will not be filtered to match the genome contigs.</p></li>
<li><p><strong>kwargs</strong> (<em>dict</em><em> , </em><em>optional</em>) – <p>Provider specific options.</p>
<dl class="simple">
<dt>toplevel<span class="classifier">bool , optional</span></dt><dd><p>Ensembl only: Always download the toplevel genome. Ignores potential primary assembly.</p>
</dd>
<dt>version<span class="classifier">int , optional</span></dt><dd><p>Ensembl only: Specify release version. Default is latest.</p>
</dd>
<dt>to_annotation<span class="classifier">text , optional</span></dt><dd><p>URL only: direct link to annotation file.
Required if this is not the same directory as the fasta.</p>
</dd>
<dt>path_to_annotation<span class="classifier">text, optional</span></dt><dd><p>Local only: path to local annotation file.
Required if this is not the same directory as the fasta.</p>
</dd>
</dl>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Genome class with the installed genome</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="genomepy.genome.Genome.html#genomepy.genome.Genome" title="genomepy.genome.Genome">Genome</a></p>
</dd>
</dl>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="genomepy.functions.head_annotations.html" class="btn btn-neutral float-left" title="genomepy.functions.head_annotations" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="genomepy.functions.list_available_genomes.html" class="btn btn-neutral float-right" title="genomepy.functions.list_available_genomes" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright Siebren Frölich, Maarten van der Sande, Tilman Schäfers and Simon van Heeringen.
      <span class="lastupdated">Last updated on 2023-05-31, 11:40 (UTC).
      </span></p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>