<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>genomepy.genome.Genome &mdash; genomepy  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="genomepy.genome.sequences" href="genomepy.genome.sequences.html" />
    <link rel="prev" title="genomepy.genome.generate_gap_bed" href="genomepy.genome.generate_gap_bed.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> genomepy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../index.html">Home</a></li>
<li class="toctree-l1"><a class="reference internal" href="../content/installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../content/command_line.html">Command line documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../content/api_core.html">Python API documentation (core)</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="genomepy.html">Python API documentation (full)</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="genomepy.annotation.html">genomepy.annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.argparse_support.html">genomepy.argparse_support</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.caching.html">genomepy.caching</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.cli.html">genomepy.cli</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.config.html">genomepy.config</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.exceptions.html">genomepy.exceptions</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.files.html">genomepy.files</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.functions.html">genomepy.functions</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="genomepy.genome.html">genomepy.genome</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="genomepy.genome.generate_fa_sizes.html">genomepy.genome.generate_fa_sizes</a></li>
<li class="toctree-l3"><a class="reference internal" href="genomepy.genome.generate_gap_bed.html">genomepy.genome.generate_gap_bed</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">genomepy.genome.Genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="genomepy.genome.sequences.html">genomepy.genome.sequences</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.online.html">genomepy.online</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.plugins.html">genomepy.plugins</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.providers.html">genomepy.providers</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.seq.html">genomepy.seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomepy.utils.html">genomepy.utils</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../content/help_faq.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="../content/about.html">About</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">genomepy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="genomepy.html">genomepy</a> &raquo;</li>
          <li><a href="genomepy.genome.html">genomepy.genome</a> &raquo;</li>
      <li>genomepy.genome.Genome</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/vanheeringen-lab/genomepy/blob/develop/docs/_autosummary/genomepy.genome.Genome.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="genomepy-genome-genome">
<h1>genomepy.genome.Genome<a class="headerlink" href="#genomepy-genome-genome" title="Permalink to this headline"></a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="genomepy.genome.Genome">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genomepy.genome.</span></span><span class="sig-name descname"><span class="pre">Genome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genomes_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomepy.genome.Genome" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">pyfaidx.Fasta</span></code></p>
<p>pyfaidx Fasta object of a genome with additional attributes &amp; methods.</p>
<p>Generates a genome index file, sizes file and gaps file of the genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – Genome name</p></li>
<li><p><strong>genomes_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Genome installation directory</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An object that provides a pygr compatible interface.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pyfaidx.Fasta</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="genomepy.genome.Genome.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genomes_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomepy.genome.Genome.__init__" title="Permalink to this definition"></a></dt>
<dd><p>An object that provides a pygr compatible interface.
filename: name of fasta file</p>
</dd></dl>

<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#genomepy.genome.Genome.__init__" title="genomepy.genome.Genome.__init__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__init__</span></code></a>(name[, genomes_dir])</p></td>
<td><p>An object that provides a pygr compatible interface.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">close</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#genomepy.genome.Genome.get_random_sequences" title="genomepy.genome.Genome.get_random_sequences"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_random_sequences</span></code></a>([n, length, chroms, ...])</p></td>
<td><p>Return random genomic sequences.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#genomepy.genome.Genome.get_seq" title="genomepy.genome.Genome.get_seq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_seq</span></code></a>(name, start, end[, rc])</p></td>
<td><p>Return a sequence by record name and interval [start, end).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#genomepy.genome.Genome.get_spliced_seq" title="genomepy.genome.Genome.get_spliced_seq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_spliced_seq</span></code></a>(name, intervals[, rc])</p></td>
<td><p>Return a sequence by record name and list of intervals</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">items</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">keys</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#genomepy.genome.Genome.track2fasta" title="genomepy.genome.Genome.track2fasta"><code class="xref py py-obj docutils literal notranslate"><span class="pre">track2fasta</span></code></a>(track[, fastafile, stranded, ...])</p></td>
<td><p>Return a list of fasta sequences as Sequence objects as directed from the track(s).</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">values</span></code>()</p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<p class="rubric">Attributes</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#genomepy.genome.Genome.gaps" title="genomepy.genome.Genome.gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gaps</span></code></a></p></td>
<td><p>contigs and the number of Ns contained</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#genomepy.genome.Genome.plugin" title="genomepy.genome.Genome.plugin"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plugin</span></code></a></p></td>
<td><p>dict of all active plugins and their properties</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#genomepy.genome.Genome.sizes" title="genomepy.genome.Genome.sizes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sizes</span></code></a></p></td>
<td><p>contigs and their lengths</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#genomepy.genome.Genome.name" title="genomepy.genome.Genome.name"><code class="xref py py-obj docutils literal notranslate"><span class="pre">name</span></code></a></p></td>
<td><p>genome name</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#genomepy.genome.Genome.genomes_dir" title="genomepy.genome.Genome.genomes_dir"><code class="xref py py-obj docutils literal notranslate"><span class="pre">genomes_dir</span></code></a></p></td>
<td><p>path to the genomepy genomes directory</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#genomepy.genome.Genome.genome_file" title="genomepy.genome.Genome.genome_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">genome_file</span></code></a></p></td>
<td><p>path to the genome fasta</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#genomepy.genome.Genome.genome_dir" title="genomepy.genome.Genome.genome_dir"><code class="xref py py-obj docutils literal notranslate"><span class="pre">genome_dir</span></code></a></p></td>
<td><p>path to the genome directory</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#genomepy.genome.Genome.index_file" title="genomepy.genome.Genome.index_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">index_file</span></code></a></p></td>
<td><p>path to the genome index</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#genomepy.genome.Genome.sizes_file" title="genomepy.genome.Genome.sizes_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sizes_file</span></code></a></p></td>
<td><p>path to the chromosome sizes file</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#genomepy.genome.Genome.gaps_file" title="genomepy.genome.Genome.gaps_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gaps_file</span></code></a></p></td>
<td><p>path to the chromosome gaps file</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#genomepy.genome.Genome.annotation_gtf_file" title="genomepy.genome.Genome.annotation_gtf_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">annotation_gtf_file</span></code></a></p></td>
<td><p>path to the gene annotation GTF file</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#genomepy.genome.Genome.annotation_bed_file" title="genomepy.genome.Genome.annotation_bed_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">annotation_bed_file</span></code></a></p></td>
<td><p>path to the gene annotation BED file</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#genomepy.genome.Genome.readme_file" title="genomepy.genome.Genome.readme_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">readme_file</span></code></a></p></td>
<td><p>path to the README file</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#genomepy.genome.Genome.tax_id" title="genomepy.genome.Genome.tax_id"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tax_id</span></code></a></p></td>
<td><p>genome taxonomy identifier</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#genomepy.genome.Genome.assembly_accession" title="genomepy.genome.Genome.assembly_accession"><code class="xref py py-obj docutils literal notranslate"><span class="pre">assembly_accession</span></code></a></p></td>
<td><p>genome assembly accession</p></td>
</tr>
</tbody>
</table>
<dl class="py attribute">
<dt class="sig sig-object py" id="genomepy.genome.Genome.annotation_bed_file">
<span class="sig-name descname"><span class="pre">annotation_bed_file</span></span><a class="headerlink" href="#genomepy.genome.Genome.annotation_bed_file" title="Permalink to this definition"></a></dt>
<dd><p>path to the gene annotation BED file</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genomepy.genome.Genome.annotation_gtf_file">
<span class="sig-name descname"><span class="pre">annotation_gtf_file</span></span><a class="headerlink" href="#genomepy.genome.Genome.annotation_gtf_file" title="Permalink to this definition"></a></dt>
<dd><p>path to the gene annotation GTF file</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genomepy.genome.Genome.assembly_accession">
<span class="sig-name descname"><span class="pre">assembly_accession</span></span><a class="headerlink" href="#genomepy.genome.Genome.assembly_accession" title="Permalink to this definition"></a></dt>
<dd><p>genome assembly accession</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genomepy.genome.Genome.gaps">
<span class="sig-name descname"><span class="pre">gaps</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#genomepy.genome.Genome.gaps" title="Permalink to this definition"></a></dt>
<dd><p>contigs and the number of Ns contained</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>contents of the gaps file</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genomepy.genome.Genome.gaps_file">
<span class="sig-name descname"><span class="pre">gaps_file</span></span><a class="headerlink" href="#genomepy.genome.Genome.gaps_file" title="Permalink to this definition"></a></dt>
<dd><p>path to the chromosome gaps file</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genomepy.genome.Genome.genome_dir">
<span class="sig-name descname"><span class="pre">genome_dir</span></span><a class="headerlink" href="#genomepy.genome.Genome.genome_dir" title="Permalink to this definition"></a></dt>
<dd><p>path to the genome directory</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genomepy.genome.Genome.genome_file">
<span class="sig-name descname"><span class="pre">genome_file</span></span><a class="headerlink" href="#genomepy.genome.Genome.genome_file" title="Permalink to this definition"></a></dt>
<dd><p>path to the genome fasta</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genomepy.genome.Genome.genomes_dir">
<span class="sig-name descname"><span class="pre">genomes_dir</span></span><a class="headerlink" href="#genomepy.genome.Genome.genomes_dir" title="Permalink to this definition"></a></dt>
<dd><p>path to the genomepy genomes directory</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomepy.genome.Genome.get_random_sequences">
<span class="sig-name descname"><span class="pre">get_random_sequences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chroms</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outtype</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'list'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomepy.genome.Genome.get_random_sequences" title="Permalink to this definition"></a></dt>
<dd><p>Return random genomic sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n</strong> (<em>int</em><em> , </em><em>optional</em>) – Number of sequences to return.</p></li>
<li><p><strong>length</strong> (<em>int</em><em> , </em><em>optional</em>) – Length of sequences to return.</p></li>
<li><p><strong>chroms</strong> (<em>list</em><em> , </em><em>optional</em>) – Return sequences only from these chromosomes.</p></li>
<li><p><strong>max_n</strong> (<em>float</em><em> , </em><em>optional</em>) – Maximum fraction of Ns.</p></li>
<li><p><strong>outtype</strong> (<em>string</em><em> , </em><em>optional</em>) – return the output as list or string.
Options: “list” or “string”, default: “list”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>coordinates as lists or strings:
List with [chrom, start, end] genomic coordinates.
String with “chrom:start-end” genomic coordinates
(can be used as input for track2fasta).</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomepy.genome.Genome.get_seq">
<span class="sig-name descname"><span class="pre">get_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomepy.genome.Genome.get_seq" title="Permalink to this definition"></a></dt>
<dd><p>Return a sequence by record name and interval [start, end).</p>
<p>Coordinates are 1-based, end-exclusive.
If rc is set, reverse complement will be returned.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomepy.genome.Genome.get_spliced_seq">
<span class="sig-name descname"><span class="pre">get_spliced_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomepy.genome.Genome.get_spliced_seq" title="Permalink to this definition"></a></dt>
<dd><p>Return a sequence by record name and list of intervals</p>
<p>Interval list is an iterable of [start, end].
Coordinates are 1-based, end-exclusive.
If rc is set, reverse complement will be returned.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genomepy.genome.Genome.index_file">
<span class="sig-name descname"><span class="pre">index_file</span></span><a class="headerlink" href="#genomepy.genome.Genome.index_file" title="Permalink to this definition"></a></dt>
<dd><p>path to the genome index</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genomepy.genome.Genome.name">
<span class="sig-name descname"><span class="pre">name</span></span><a class="headerlink" href="#genomepy.genome.Genome.name" title="Permalink to this definition"></a></dt>
<dd><p>genome name</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genomepy.genome.Genome.plugin">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">plugin</span></span><a class="headerlink" href="#genomepy.genome.Genome.plugin" title="Permalink to this definition"></a></dt>
<dd><p>dict of all active plugins and their properties</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genomepy.genome.Genome.readme_file">
<span class="sig-name descname"><span class="pre">readme_file</span></span><a class="headerlink" href="#genomepy.genome.Genome.readme_file" title="Permalink to this definition"></a></dt>
<dd><p>path to the README file</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genomepy.genome.Genome.sizes">
<span class="sig-name descname"><span class="pre">sizes</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#genomepy.genome.Genome.sizes" title="Permalink to this definition"></a></dt>
<dd><p>contigs and their lengths</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>contents of the sizes file</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genomepy.genome.Genome.sizes_file">
<span class="sig-name descname"><span class="pre">sizes_file</span></span><a class="headerlink" href="#genomepy.genome.Genome.sizes_file" title="Permalink to this definition"></a></dt>
<dd><p>path to the chromosome sizes file</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genomepy.genome.Genome.tax_id">
<span class="sig-name descname"><span class="pre">tax_id</span></span><a class="headerlink" href="#genomepy.genome.Genome.tax_id" title="Permalink to this definition"></a></dt>
<dd><p>genome taxonomy identifier</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomepy.genome.Genome.track2fasta">
<span class="sig-name descname"><span class="pre">track2fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">track</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fastafile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stranded</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extend_up</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extend_down</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomepy.genome.Genome.track2fasta" title="Permalink to this definition"></a></dt>
<dd><p>Return a list of fasta sequences as Sequence objects
as directed from the track(s).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>track</strong> (<em>list/region file/bed file</em>) – region(s) you wish to translate to fasta.
Example input files can be found in genomepy/tests/data/regions.*</p></li>
<li><p><strong>fastafile</strong> (<em>bool</em><em> , </em><em>optional</em>) – return Sequences as list or save to file? (default: list)</p></li>
<li><p><strong>stranded</strong> (<em>bool</em><em> , </em><em>optional</em>) – return sequences for both strands? Required BED6 (or higher) as input (default: False)</p></li>
<li><p><strong>extend_up</strong> (<em>int</em><em> , </em><em>optional</em>) – extend the sequences up? (command is strand sensitive, default: 0)</p></li>
<li><p><strong>extend_down</strong> (<em>int</em><em> , </em><em>optional</em>) – extend the sequences down? (command is strand sensitive, default: 0)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="genomepy.genome.generate_gap_bed.html" class="btn btn-neutral float-left" title="genomepy.genome.generate_gap_bed" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="genomepy.genome.sequences.html" class="btn btn-neutral float-right" title="genomepy.genome.sequences" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright Simon van Heeringen, Siebren Frölich, Maarten van der Sande.
      <span class="lastupdated">Last updated on 2021-11-18, 10:10 (UTC).
      </span></p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>